File: Trees/preorder_stack.cpp
#include <iostream>
#include<stack>
using namespace std;
struct Node{
    int key;
    Node* left;
    Node* right;

    Node(int x){
        key = x;
        left = NULL;
        right = NULL;
    }
};

void preorder(Node* root){
    stack <Node* > s;
    s.push(root);

    while(!s.empty()){
        Node* curr_node = s.top();
        cout << curr_node->key<< " ";
        s.pop();

        if (curr_node->right != NULL){
            s.push(curr_node->right);
        }
        if (curr_node->left !=NULL){
            s.push(curr_node->left);
        }

    }


}

int main(){
    struct Node* root = new Node(10);
    root->left = new Node(8);
    root->right = new Node(2);
    root->left->left = new Node(3);
    root->left->right = new Node(5);
    root->right->left = new Node(2);
    preorder(root);
}

File: Trees/preorder_stack.py
class Node:
    def __init__(self,key):
        self.key = key
        self.right = None
        self.left = None

'''
we use a stack and push the right node first because we have to pop the left child
first for the preorder
'''    
def preorder(root):
    s = []
    s.append(root)

    while s:
        curr_node = s.pop()
        print(curr_node.key,end = ' ')
        if curr_node.right is not None:
            s.append(curr_node.right)
        if curr_node.left is not None:
            s.append(curr_node.left)

root = Node(10) 
root.left = Node(8) 
root.right = Node(2) 
root.left.left = Node(3) 
root.left.right = Node(5) 
root.right.left = Node(2) 
preorder(root) 
